src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -91,46 +70,83 @@ public BoundLambda(SyntaxNode syntax, BoundBlock body, ImmutableArray<Diagnostic
-
-                foreach (var info in _inferredReturnTypeUseSiteDiagnostics)
+                foreach (var info in InferredReturnType.UseSiteDiagnostics)
                 {
                     useSiteDiagnostics.Add(info);
                 }
             }
-
-            return _inferredReturnType;
+            if (nullableState == null)
+            {
+                return InferredReturnType.Type;
+            }
+            else
+            {
+                var builder = ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)>.GetInstance();
+                // Diagnostics from NullableWalker.Analyze can be dropped here since Analyze
+                // will be called again from NullableWalker.ApplyConversion when the
+                // BoundLambda is converted to an anonymous function.
+                var diagnostics = DiagnosticBag.GetInstance();
+                var delegateType = Type.GetDelegateType();
+                var compilation = Binder.Compilation;
+                var conversions = Binder.Conversions.WithNullability(includeNullability: true);
+                NullableWalker.Analyze(compilation, this, diagnostics, delegateInvokeMethod: delegateType?.DelegateInvokeMethod, returnTypes: builder, initialState: nullableState);
+                diagnostics.Free();
+                var inferredReturnType = InferReturnType(builder, compilation, conversions, delegateType, Symbol.IsAsync);
+                builder.Free();
+                return inferredReturnType.Type;
+            }
         }
 
+        /// <summary>
+        /// Indicates the type of return statement with no expression. Used in InferReturnType.
+        /// </summary>
+        internal static readonly TypeSymbolWithAnnotations NoReturnExpression = TypeSymbolWithAnnotations.Create(new UnsupportedMetadataTypeSymbol());
+
         /// <summary>
         /// Behavior of this function should be kept aligned with <see cref="UnboundLambdaState.ReturnInferenceCacheKey"/>.
         /// </summary>
-        private static TypeSymbolWithAnnotations InferReturnType(
-            BoundBlock block,
-            Binder binder,
+        internal static InferredLambdaReturnType InferReturnType(ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)> builder, CSharpCompilation compilation, Conversions conversions, TypeSymbol delegateType, bool isAsync)
+        {
+            var types = new HashSet<TypeSymbolWithAnnotations>(TypeSymbolWithAnnotations.EqualsComparer.Instance);
+            bool hasReturnWithoutArgument = false;
+            RefKind refKind = RefKind.None;
+            foreach (var (rk, type) in builder)
+            {
+                if (rk != RefKind.None)
+                {
+                    refKind = rk;
+                }
+                if ((object)type == NoReturnExpression)
+                {
+                    hasReturnWithoutArgument = true;
+                }
+                else
+                {
+                    types.Add(type);
+                }
```

Comment:
Is it possible to get here with a lambda that has both kinds of returns (with and without expressions)?

Reply:
`F(() => { if (true) return 1; return; });`

---
In reply to: [192904858](https://github.com/dotnet/roslyn/pull/27283#discussion_r192904858) [](ancestors = 192904858)